migration-engine/core/src/migration/datamodel_differ/fields.rs

```diff
@@ -0,0 +1,48 @@
+use super::directives::{directives_match, DirectiveDiffer};
+use datamodel::ast;
+
+/// Implements the logic to diff a pair of [Field ASTs](/datamodel/ast/struct.Field.html).
+#[derive(Debug)]
+pub(crate) struct FieldDiffer<'a> {
+    pub(crate) previous: &'a ast::Field,
+    pub(crate) next: &'a ast::Field,
+}
+
+impl<'a> FieldDiffer<'a> {
+    fn previous_directives(&self) -> impl Iterator<Item = &ast::Directive> {
+        self.previous.directives.iter()
+    }
+
+    fn next_directives(&self) -> impl Iterator<Item = &ast::Directive> {
+        self.next.directives.iter()
+    }
+
+    /// Iterator over the directives present in `next` but not in `previous`.
+    pub(crate) fn created_directives(&self) -> impl Iterator<Item = &ast::Directive> {
+        self.next_directives().filter(move |next_directive| {
+            self.previous_directives()
+                .find(|previous_directive| directives_match(previous_directive, next_directive))
+                .is_none()
+        })
+    }
+
+    /// Iterator over the directives present in `previous` but not in `next`.
+    pub(crate) fn deleted_directives(&self) -> impl Iterator<Item = &ast::Directive> {
```

Comment:
I can't yet wrap my head around it but it looks like there's some duplication for all things that have directives. Would be it be feasible to dry that up or would that complicate the code too much?

Reply:
I thought about a trait implemented by EnumDiffer, FieldDiffer and ModelDiffer. The issue is with the "repeated directives" complication on models - either we implement the methods separately for Models, or we make the trait more complicated than necessary for fields and enums. Maybe there's a clean way to abstract this, I'll try to come up with something.